app/src/main/java/com/google/android/material/tabs/TabLayoutMediator.java

```diff
@@ -0,0 +1,309 @@
+/*
+ * Copyright 2018 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.android.material.tabs;
+
+import static androidx.viewpager2.widget.ViewPager2.SCROLL_STATE_DRAGGING;
+import static androidx.viewpager2.widget.ViewPager2.SCROLL_STATE_IDLE;
+import static androidx.viewpager2.widget.ViewPager2.SCROLL_STATE_SETTLING;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.recyclerview.widget.RecyclerView;
+import androidx.viewpager2.widget.ViewPager2;
+import java.lang.ref.WeakReference;
+
+/**
+ * A mediator to link a TabLayout with a ViewPager2. The mediator will synchronize the ViewPager2's
+ * position with the selected tab when a tab is selected, and the TabLayout's scroll position when
+ * the user drags the ViewPager2. TabLayoutMediator will listen to ViewPager2's OnPageChangeCallback
+ * to adjust tab when ViewPager2 moves. TabLayoutMediator listens to TabLayout's
+ * OnTabSelectedListener to adjust VP2 when tab moves. TabLayoutMediator listens to RecyclerView's
+ * AdapterDataObserver to recreate tab content when dataset changes.
+ *
+ * <p>Establish the link by creating an instance of this class, make sure the ViewPager2 has an
+ * adapter and then call {@link #attach()} on it. Instantiating a TabLayoutMediator will only create
+ * the mediator object, {@link #attach()} will link the TabLayout and the ViewPager2 together. When
+ * creating an instance of this class, you must supply an implementation of {@link
+ * TabConfigurationStrategy} in which you set the text of the tab, and/or perform any styling of the
+ * tabs that you require. Changing ViewPager2's adapter will require a {@link #detach()} followed by
+ * {@link #attach()} call. Changing the ViewPager2 or TabLayout will require a new instantiation of
+ * TabLayoutMediator.
+ */
+public final class TabLayoutMediator {
```

Comment:
> Hmm. I want to keep the amount of code that we need to take care of as small as possible. At some point, we have to upgrade the dependencies. We already use 64,945 out of 65,536 methods, so we can not go much longer without multidex, anyway.

So what do you suggest? It isn't a fork, can (and should) be deleted after the upgrade.

Reply:
> I will play around with ProGuard to see if we can get a bit more room for new methods without enabling MultiDex. I mean, even big apps like WhatsApp work without MultiDex. Something seems to waste methods...

They may rely more on native codes I believe. I will also check what has consumed the methods number.

Reply:
I brought it down to around 44k by removing `-keep class androidx.** { *; }` from the proguard config. Currently fine-tuning the proguard settings so that the tests pass.

Reply:
Let me know when you landed it. Thanks!

Reply:
Done :) #4035 is merged, so we have a bit more room for library updates now